Specify the tokenizer for window.open's features argument

zcorpan opened the pull request:
This was specified in CSSOM View but the "noopener" feature did
not use the same tokenizer as the legacy features. Fixes #2474.

Also specify the aliases screenx, screeny, innerwidth, innerheight
for left, top, width, and height, respectively. Part of #2464.
Closes https://github.com/w3c/csswg-drafts/pull/1128.

---

This needs updates in cssom-view as well.

Maybe the aliases should be a separate commit... or maybe they should be in cssom-view still, not sure.

annevk commented on source:
```diff
@@ -78739,6 +78741,66 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+
+   <li>
+    <p>For each <var>token</var> of <var>tokens</var>:</p>
+    <ol>
```
Newline before `<ol>`.

annevk commented on source:
```diff
@@ -78739,6 +78741,66 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+   <li><p>Let <var>map</var> be a new <span>ordered map</span>.</p></li>
+
+   <li>
+    <p>For each <var>token</var> of <var>tokens</var>:</p>
```
Link "For each".

annevk commented on source:
```diff
@@ -78739,6 +78741,66 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+
+     <li><p>Let <var>position</var> point at the first character of <var>input</var>.</p></li>
+
+     <li><p><span>Skip ASCII whitespace</span>.
```
You need to pass this parameters these days.

annevk commented on source:
```diff
@@ -78739,6 +78741,66 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+   <li>
+    <p>For each <var>token</var> of <var>tokens</var>:</p>
+    <ol>
+     <li><p>Let <var>input</var> be <var>token</var>.
```
Is this needed?

annevk commented on source:
```diff
@@ -78739,6 +78741,66 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+
+     <li><p><span>Skip ASCII whitespace</span>.
+
+     <li><p><span>Collect a sequence of code points</span> that are not <span>ASCII
```
This also needs to be passed more arguments.

annevk commented on source:
```diff
@@ -78739,6 +78741,66 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+     <var>name</var>.</p></li>
+
+     <li><p>If <var>tokenizedFeatures</var>[<var>name</var>] <span data-x="map
+     exists">exists</span>, then continue.</p></li>
```
So you can't overwrite existing values?

zcorpan replied:
Right. This matches Gecko but not WebKit or Chromium.

http://software.hixie.ch/utilities/js/live-dom-viewer/saved/4986

What does Edge do?

annevk replied:
Edge overwrites, but does not appear to support left (I ended up overwriting width).

annevk commented on source:
```diff
@@ -78739,6 +78741,66 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+     <li><p>If <var>tokenizedFeatures</var>[<var>name</var>] <span data-x="map
+     exists">exists</span>, then continue.</p></li>
+
+     <li><p><span>Skip ASCII whitespace</span>.</p></li>
```
Missing arguments again.

zcorpan commented on source:
```diff
@@ -78739,6 +78741,66 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+     whitespace</span> nor U+003D (=). Let <var>name</var> be the collected characters,
+     <span>converted to ASCII lowercase</span>.</p></li>
+
+     <li><p>Set <var>name</var> to the result of <span>normalizing the feature name</span>
```
Hmm maybe the name shouldn't be normalized during tokenization.

http://software.hixie.ch/utilities/js/live-dom-viewer/saved/4987

Gecko picks the "left" value in both cases. Chromium and WebKit pick the last one specified.

http://software.hixie.ch/utilities/js/live-dom-viewer/saved/4988

Here Chromium and WebKit still pick the last one specified and use 0. Gecko has different result for the two cases...

annevk commented on source:
```diff
@@ -78739,6 +78741,66 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+   <li><p>Let <var>tokenizedFeatures</var> be a new <span>ordered map</span>.</p></li>
+
+   <li>
+    <p>For each <var>token</var> of <var>tokens</var>:</p>
```
"For each" should be a link to https://infra.spec.whatwg.org/#list-iterate

lyzadanger commented:
@zcorpan I was thinking making test coverage better in this area of the spec (7.3.1) — I'm not sure how the process typically works but would you like me to help out with writing tests for this?

zcorpan commented:
@lyzadanger that would be awesome! Thanks!

I think this could be automated by checking https://drafts.csswg.org/cssom-view/#dom-window-screenx et al. Although the features for window.open needs to be defined in terms of https://drafts.csswg.org/cssom-view/#web-exposed-screen-area as well so they match up (and to fix the privacy leak for window.open).

Another problem is that doing anything with *features* is optional (other than `noopener`), but I guess we can add metadata to the tests to that effect.

lyzadanger commented:
@zcorpan Sounds good. Just finishing up some travel; will get on this tomorrow (Friday).

zcorpan commented on source:
```diff
@@ -78739,6 +78742,66 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+     <var>position</var>.</p></li>
+
+     <li><p>If <var>position</var> is not past the end of <var>token</var> and the character at
+     <var>position</var> is U+003D (=), then advance <var>position</var> by 1.</p></li>
```
Hmm there is a bug here. If it's not past the end and the character is not `=`, we shouldn't collect a value. Maybe it would be better to split on the first `=` and then trim leading and trailing whitespace from name and value?

zcorpan commented:
I looked at [the code in Chromium](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/page/WindowFeatures.cpp?q=windowfeatures+package:%5Echromium$&l=1) and reverse-engineered Edge. It appears there is almost perfect interop between Edge and Chromium and WebKit in how to tokenize. The only difference seems to be handling of U+0000 (Edge cuts the value at the first U+0000, Chromium treats it as a separator like space, WebKit treats it as a non-separator like x). I'll specify Chromium's tokenizer.

For example `==width==450==height==450` "works".

annevk commented:
What does Chrome do for U+0001? I don't see why we'd treat U+0000 as a separator.

zcorpan commented:
It's a non-separator. Happy to spec WebKit-like instead on that point.

Another interesting bit is U+000C, which only Gecko treats as a separator character. Likely it is Web-compatible to treat it as a separator so this is more consistent with everything else.

zcorpan commented:
I found a case that is not interoperable between Edge and WebKit/Chromium, see https://github.com/whatwg/html/pull/2476/commits/ca2aa5d8dfc221b3f061f2a32c53b5ba61faecea

Implementation of the proposed spec with a few tests: http://software.hixie.ch/utilities/js/live-dom-viewer/saved/4998

lyzadanger commented on source:
```diff
@@ -78739,6 +78742,104 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+
+      <ol>
+       <li><p>If the code point at <var>position</var> in <var>features</var> is U+002C (,), or if
+       it is not a <span>window features separator</span>, then break.</p></li>
```
Does this _break_ to 

> 3. While position is not past the end of features:

(line 78754)?

zcorpan replied:
Yes, per https://infra.spec.whatwg.org/#iteration-break -- @annevk @domenic should we xref these terms?

zcorpan replied:
I mean no. It breaks the nearest iteration.

lyzadanger replied:
Which is

> While position is not past the end of features and the code point at position in features is not U+003D (=):

??

In that case, does this imply that any `','` characters between the `name` and the first instance of `'='` are ignored? e.g. the case:

`'noopener ,=,'`

How does that tokenize?

zcorpan replied:
```
log: noopener ,=foo,
log: {
  "noopener": "",
  "foo": ""
}
```
I could step through the spec but maybe it's simpler to step through the JS implementation with some browser's devtools? :-)

lyzadanger replied:
Ah, yes, good idea (I'm new to the part of this that involves prototyping implementation via JS). I'll walk through that more carefully and do some of my own tests against it there. In the end, I want to be sure that the spec language is in accordance to the implementation (obviously :) ). More soon.

lyzadanger replied:
tl;dr I think things are fine as-is.

My confusion was that, at this point in the spec, we are nested a few levels deep, but if I evaluate those levels, it's `while (a) > while (b) > if`. The `break` statement in that `if` conditional will return control to `while (a)`, which is as it should be, but I had to think more programmatically than I was initially. 👍  and thanks for your patience.

domenic replied:
IMO yes we should xref those terms. We're taking an incremental approach to introducing infra into HTML, so it's good to use this opportunity to introduce it here.

lyzadanger commented on source:
```diff
@@ -78739,6 +78742,104 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+       </li>
+
+       <li><p><span>Collect a sequence of code points</span> that are not <span data-x="window
+       features separator">window features separators</span> code points from <var>features</var>
```
What happens to anything after the last non-separator code point in `value`? e.g. in the case `'noopener=yes yes,foo=bar'`, I believe you end up with `('noopener', 'yes')` for the first tokenized feature, but what happens to the other `'yes'`? It seems like this might put you into the next loop of tokenization with some remaining characters before the next `','`?

lyzadanger replied:
...or does this tokenize to:

* name `'noopener'`, value `'yes'`
* name `'yes'`, value `''`
* name `'foo'`, value `'bar'`

zcorpan replied:
Yes. Added that case in http://software.hixie.ch/utilities/js/live-dom-viewer/saved/5001 to confirm. (Also fixed a typo `asciiLowercase` -> `asciiLowerCase` from 4998.)

zcorpan commented:
I've cleaned up the history and rebased to resolve conflicts. I left the like-Edge change as a separate commit to make it clearer what the difference is from like-Chromium and -WebKit.

annevk commented:
When you land you'll squash these I assume? (The second commit doesn't seem to have enough context on its own.)

zcorpan commented:
Yes

zcorpan commented:
Tests for `noopener` are (almost) ready: https://github.com/w3c/web-platform-tests/issues/5306

@annevk can you do another round of editorial review? It's possible I made a mistake when rewriting history and resolving conflicts. I'll file browser bugs tomorrow.

annevk commented on source:
```diff
@@ -78733,6 +78739,104 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+  </ol>
+
+  <p>A code point is a <dfn>window features separator</dfn> if it is one of <span>ASCII
+  whitespace</span>, U+003D (=), or U+002C (,).</p>
```
s/one of//

zcorpan commented:
Fixed nit and renamed to "feature separator"

zcorpan commented:
https://bugs.webkit.org/show_bug.cgi?id=170548
https://bugs.chromium.org/p/chromium/issues/detail?id=709037
https://bugzilla.mozilla.org/show_bug.cgi?id=1354174
https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/11544316/
